runtime/common/src/migrations.rs

```diff
@@ -564,6 +573,105 @@ where
             Box::new(migrate_registrar_pending_verification),
             Box::new(migrate_registrar_manager),
             Box::new(migrate_pallet_xcm_v4),
+            Box::new(foreign_asset_creator_migration),
         ]
     }
 }
+
+pub struct ForeignAssetCreatorMigration<Runtime>(pub PhantomData<Runtime>);
+
+impl<Runtime> Migration for ForeignAssetCreatorMigration<Runtime>
+where
+    Runtime: pallet_foreign_asset_creator::Config,
+    <Runtime as pallet_foreign_asset_creator::Config>::ForeignAsset:
+        TryFrom<staging_xcm::v3::MultiLocation>,
+{
+    fn friendly_name(&self) -> &str {
+        "TM_ForeignAssetCreatorMigration"
+    }
+
+    fn migrate(&self, _available_weight: Weight) -> Weight {
+        use frame_support::pallet_prelude::*;
+
+        use staging_xcm::v3::MultiLocation as OldLocation;
+
+        let pallet_prefix = AssetIdToForeignAsset::<Runtime>::pallet_prefix();
+        let asset_id_to_foreign_asset_storage_prefix =
+            AssetIdToForeignAsset::<Runtime>::storage_prefix();
+        let foreign_asset_to_asset_id_prefix = ForeignAssetToAssetId::<Runtime>::storage_prefix();
+
+        // Data required to migrate ForeignAsset values
+        // Read all the data into memory.
+        let asset_id_to_foreign_asset_data: Vec<_> =
+            storage_key_iter::<AssetId<Runtime>, OldLocation, Blake2_128Concat>(
+                pallet_prefix,
+                asset_id_to_foreign_asset_storage_prefix,
+            )
+            .drain()
+            .collect();
+
+        // Data required to migrate ForeignAsset keys
+        let foreign_asset_to_asset_id_data: Vec<_> =
+            storage_key_iter::<OldLocation, AssetId<Runtime>, Blake2_128Concat>(
+                pallet_prefix,
+                foreign_asset_to_asset_id_prefix,
+            )
+            .drain()
+            .collect();
+
+        let foreign_asset_to_asset_id_count = foreign_asset_to_asset_id_data.len();
+
+        let migrated_count = asset_id_to_foreign_asset_data
+            .len()
+            .saturating_add(foreign_asset_to_asset_id_data.len());
+
+        log::info!("Migrating {:?} elements", migrated_count);
+
+        // Removing older entries with old location as key
+        let mut removal_result = storage::unhashed::clear_prefix(
+            &ForeignAssetToAssetId::<Runtime>::final_prefix(),
+            None,
+            None,
+        );
+        while removal_result.maybe_cursor.is_some() {
+            removal_result = storage::unhashed::clear_prefix(
+                &ForeignAssetToAssetId::<Runtime>::final_prefix(),
+                None,
+                removal_result.maybe_cursor.as_deref(),
+            );
+        }
```

Comment:
This may not be needed because `storage_key_iter().drain()` above already removes from storage

Reply:
Oh. I did not know that drain also removes the item from storage.

Reply:
Me neither, I had to look at docs: https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/struct.PrefixIterator.html#method.drain

Reply:
Removed this call in next commit.